### PR TITLE
Feature/861 stan threads always on

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submisison Checklist

- [x] Run tests: `./runCmdStanTests.py src/test`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Add logic to CmdStan makefile so that makefile variable `STAN_THREADS` is set to `TRUE`.

#### Intended Effect:

Avoid linker errors.  see issue #861 

#### How to Verify:

Unit tests

#### Side Effects:

Some models may run slower, see discussion here:  https://discourse.mc-stan.org/t/makefiles-and-compiler-flag-dstan-threads-always-on/14482/5?u=mitzimorris

#### Documentation:

NA - compiler flag `STAN_THREADS` no mentioned in user's guide.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Columbia University
 
By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by wds15** [comment]
Does this qualify for 2.23? Reduce sum will make people want to use this.

File: makefile [hunk 1]
```diff
@@ -31,6 +32,10 @@ O_STANC ?= 0
 INC_FIRST ?= -I src -I $(STAN)src -I $(RAPIDJSON)
 USER_HEADER ?= $(dir $<)user_header.hpp
 
+ifeq ($(origin STAN_THREADS), undefined)
```
Comment (rok-cesnovar) [inline makefile:35]: Could this just be `STAN_THREADS ?= TRUE`

**Comment by rok-cesnovar** [comment]
> Does this qualify for 2.23? Reduce sum will make people want to use this.

Given that reduce_sum is the by far biggest new feature of 2.23 I would be for merging this.

File: makefile (hunk 1 above)
Comment (wds15) [reply makefile:35]: I would think the same. `STAN_THREADS ?= true` should do it.

**Comment by wds15** [comment]
I just restarted the tests as the performance tests are counter to what I recall.

I am good with making STAN_THREADS on by default; it's just that seeing the performance results makes me cautious.

**Comment by rok-cesnovar** [comment]
I ran longer tests:
Linux - 
https://jenkins.mc-stan.org/job/CmdStan%20Performance%20Tests/job/Custom/153/artifact/linux_compare_results.xml/*view*/

Mac - https://jenkins.mc-stan.org/job/CmdStan%20Performance%20Tests/job/Custom/153/artifact/macos_compare_results.xml/*view*/

Seems worse than we thought it is.

**Comment by wds15** [comment]
Yeah... I am coming to the same conclusion. I wonder if optimisation which we have done in the past releases were more beneficial for non-threading code. Is it possible to see over time (or by release) the performance of individual models? I am just curious.

In any case, I don't think that we can just switch to threads being on for the moment being. In particular, it should not be done as a last minute decision for the release.

Also: I think we should be able to have an even better setup:

- we build the pre-build stuff with and without threads
- if the user does not define STAN_THREADS, then we simply auto-detect it from the model. That is relatively easy for the moment since this is merely a grep for "map_rect" or "reduce_sum". If these are found in the model, then you get threading otherwise not.
- Whenever the user defines STAN_THREADS, then he gets what he asks for

**Comment by mitzimorris** [comment]
> Also: I think we should be able to have an even better setup:
> 
> * we build the pre-build stuff with and without threads
> * if the user does not define STAN_THREADS, then we simply auto-detect it from the model. That is relatively easy for the moment since this is merely a grep for "map_rect" or "reduce_sum". If these are found in the model, then you get threading otherwise not.
> * Whenever the user defines STAN_THREADS, then he gets what he asks for

interesting idea - I'll investigate.

**Comment by mitzimorris** [comment]
just to clarify - let's not merge this PR and I will continue to work on logic.

**Comment by wds15** [comment]
If you want I can give it a try... but give me some days for it.

**Pull request closed without merging**